### QUESTÃO 1: Dados os pontos $(0, 0), (0.2, 0.60667), (0.3, 0.92252), (0.5, 1.60443), (0.6, 1.98054), (0.8, 2.82936), (1.0, 3.84147), (1.4, 6.52945)$ e $(1.5, 7.37249)$, faça o seguinte:

#### a) Calcule $f(0.32)$ por interpolação linear e estime o erro.

O intervalo deve ser $[0.3, 0.5]$
$$
y = ax + b \\
0.92252 = 0.3a + b \\
1.60443 = 0.5a + b
$$

In [1]:
(1.60443 - 0.92252) / 0.2

3.40955

Então $a = 3.40955$

In [2]:
1.60443 - 0.5 * 3.40955

-0.1003449999999999

E $b = -0.100345$

Então a função de interpolação linear é:
$$
p(x) = 3.40955x - 0.100345
$$

In [3]:
def p(x): # função de interpolação acima
    return 3.40955 * x - 0.100345

p(0.32)

0.990711

$$
p(0.32) = 0.990711
$$

#### b) Calcule $f(1.46)$ por interpolação linear e estime o erro.

Usando o intervalo $[1.4, 1.5]$ e agora resolvendo pela fórmula:

In [4]:
def interpolacao_linear(a, fa, b, fb, x):
    """
    a e b são os intervalos, fa e fb são os valores da função em a e b,
    x é o ponto em que se quer descobrir o valor da função
    """
    
    return fa + ((fb-fa)/(b-a)) * (x-a)

In [5]:
interpolacao_linear(1.4, 6.52945, 1.5, 7.37249, 1.46)

7.035274

$$
p(1.46) = 7.035274
$$

#### c) Encontre um valor aproximado para $f(0.7)$ usando um polinômio interpolador de grau 2 adequado e estime o erro.

Usando os pontos $(0.5, 1.60443), (0.6, 1.98054), (0.8, 2.82936)$

Pelo polinômio de Lagrange:
$$
p(x) = 1.60443 \cdot \frac{x-0.6}{0.5-0.6} \cdot \frac{x-0.8}{0.5-0.8} + 1.98054 \cdot \frac{x-0.5}{0.6-0.5} \cdot \frac{x-0.8}{0.6-0.8} + 2.82936 \cdot \frac{x-0.5}{0.8-0.5} \cdot \frac{x-0.6}{0.8-0.6} \\
p(x) = 1.60443 \cdot \frac{x-0.6}{-0.1} \cdot \frac{x-0.8}{-0.3} + 1.98054 \cdot \frac{x-0.5}{0.1} \cdot \frac{x-0.8}{-0.2} + 2.82936 \cdot \frac{x-0.5}{0.3} \cdot \frac{x-0.6}{0.2} \\
p(x) = 1.60443 \cdot \frac{x-0.6}{0.1} \cdot \frac{x-0.8}{0.3} - 1.98054 \cdot \frac{x-0.5}{0.1} \cdot \frac{x-0.8}{0.2} + 2.82936 \cdot \frac{x-0.5}{0.3} \cdot \frac{x-0.6}{0.2} \\
p(x) = 53.481 x^2 - 74.8734 x + 25.6709 - 99.027 x^2 + 128.735 x - 39.6108 + 47.156 x^2 - 51.8716 x + 14.1468 \\
p(x) = 1.61 x^2 + 1.99 x + 0.2069
$$

In [6]:
(lambda x: 1.61*x**2 + 1.99*x + 0.2069)(0.7)

2.3888

$$
p(0.7) = 2.3888
$$

#### d) Sabendo que $f(x) = x^3 + 2x + \mathrm{sen}(x)$, encontre novas estimativas para os erros acima.

Os limites dos erros serão encontrados pelo resto de Lagrange:

$$
R(x) \leq \frac{M \cdot (x_k-x_0)^{k+1}}{(k+1)!}
$$

Onde $M$ é o valor máximo de $|f^{(k+1)}(x)|$ no intervalo $x_0 \leqslant x \leqslant x_k$ e $k$ é o grau da interpolação.

##### $f(0.32)$

Pontos: $(0.3, 0.92252), (0.5, 1.60443)$

$$
f^{(2)}(x) = 6x - \mathrm{sen}(x) \\
$$
E o valor máximo no intervalo $[0.3, 0.5]$ será em $x = 0.5$
$$
M = 6(0.5) - \mathrm{sen}(0.5) \\
M = 2.52057
$$

Então
$$
R(x) \leq \frac{M \cdot (x_k-x_0)^{k+1}}{(k+1)!} \\
R(x) \leq \frac{2.52057 \cdot (0.5-0.3)^{1+1}}{(1+1)!} \\
R(x) \leq \frac{2.52057 \cdot (0.2)^{2}}{2!} \\
R(x) \leq 0.0504114
$$

##### $f(1.46)$

Pontos: $(1.4, 6.52945), (1.5, 7.37249)$

$$
f^{(2)}(x) = 6x - \mathrm{sen}(x) \\
$$
E o valor máximo no intervalo $[1.4, 1.5]$ será em $x = 1.5$
$$
M = 6(1.5) - \mathrm{sen}(1.5) \\
M = 8.00251
$$

Então
$$
R(x) \leq \frac{M \cdot (x_k-x_0)^{k+1}}{(k+1)!} \\
R(x) \leq \frac{8.00251 \cdot (1.5-1.4)^{1+1}}{(1+1)!} \\
R(x) \leq \frac{8.00251 \cdot (0.1)^{2}}{2!} \\
R(x) \leq 0.04001255
$$

##### $f(0.7)$

Pontos: $(0.5, 1.60443), (0.6, 1.98054), (0.8, 2.82936)$

$$
f^{(3)}(x) = 6 - \cos(x) \\
$$
E o valor máximo no intervalo $[0.5, 0.8]$ será em $x = 0.8$
$$
M = 6 - \cos(0.8) \\
M = 5.30329
$$

Então
$$
R(x) \leq \frac{M \cdot (x_k-x_0)^{k+1}}{(k+1)!} \\
R(x) \leq \frac{5.30329 \cdot (0.8-0.5)^{3+1}}{(3+1)!} \\
R(x) \leq \frac{5.30329 \cdot (0.3)^{4}}{4!} \\
R(x) \leq 0.001789860375
$$

---

### QUESTÃO 3: A função $y = e^x$ é tabelada para $x \in [0, 1]$, com $h = 0.01$. Encontre o erro máximo para uma interpolação linear.

Pelo Teorema de Rolle, o erro de uma interpolação linear tem o seguinte limite
$$
E \leq \frac{M \cdot (x_1 - x_0)^2}{8}
$$
Onde $M$ é o valor máximo de $|f^{(2)}(x)|$ no intervalo $x_0 \leq x \leq x_1$.

Como qualquer derivada de $e^x$ é $e^x$, $M$ será máximo em $x = 1$

$$
M = e
$$

Então
$$
E \leq \frac{M \cdot (x_1 - x_0)^2}{8} \\
E \leq \frac{e \cdot (1 - 0)^2}{8} \\
E \leq \frac{e}{8} \\
E \leq 0.339785
$$